transport/buffer.go

```diff
@@ -0,0 +1,158 @@
+package transport
+
+import (
+	"bytes"
+	"fmt"
+	"io"
+	"sync"
+	"time"
+)
+
+// A TimestampWriterTo extends io.WriterTo with a method that returns the timestamp of when the
+// contents were last successfully written.
+type TimestampWriterTo interface {
+	io.WriterTo
+	Timestamp() time.Time
+}
+
+// Buffer is a write safe io.Writer & io.WriterTo that is used to buffer output and safely copy it
+// to a transport writer. If copying fails, buffer does not lose data.
+type Buffer struct {
+	Encoder
+
+	Metadata    Metadata
+	MaxIdleTime time.Duration
+
+	mu        sync.RWMutex
+	output    *bytes.Buffer
+	results   []Result
+	timestamp time.Time
+}
+
+// WriteResult writes structured task execution output to the response.
+func (b *Buffer) WriteResult(result Result) {
+	b.mu.Lock()
+	b.results = append(b.results, result)
+	b.mu.Unlock()
+}
+
+// Write buffers the provided data until it is consumed by a transport. It is safe for
+// concurrent use.
+func (b *Buffer) Write(data []byte) (int, error) {
+	b.mu.Lock()
+	defer b.mu.Unlock()
+
+	if b.output == nil {
+		b.output = bytes.NewBuffer(data)
+		return len(data), nil
+	}
+
+	return b.output.Write(data)
+}
+
+// WriteTo implements io.WriterTo, allowing Buffer to be used by io.Copy. It will write it's
+// contents to the provided writer. It will preserve any data in the case of error.
+func (b *Buffer) WriteTo(w io.Writer) (int64, error) {
+	output := b.copyOutput()
+	results := b.copyResults()
+
+	if len(output) <= 0 && len(results) <= 0 && time.Since(b.Timestamp()) < b.MaxIdleTime {
+		return 0, nil
+	}
+
+	resp := Response{
+		Metadata: b.Metadata,
+		Log:      output,
+		Results:  results,
+	}
+	data, err := b.Encode(resp)
+	if err != nil {
+		b.restore(results, output)
+		return 0, fmt.Errorf("failed to encode response data: %w", err)
+	}
+
+	n, err := w.Write(data)
+	if err != nil {
+		b.restore(results, output)
```

Comment:
**[golangci-lint]** <sub>reported by [reviewdog](https://github.com/reviewdog/reviewdog) :dog:</sub>
Error return value of `b.restore` is not checked (errcheck)